### ISSUE TITLE
Add support to hook into require

### ISSUE THREAD

**Issue by Papipo** [tv-labs/lua#34]
Hi!

I have this app in which users can provide their lua scripts. I don't have those in the filesystem so I need to intercept `require` calls somehow (I will be returning them from the database or as chunks from ETS).

An option is to override `require` itself, but I think that adding a `searcher` after the preload one is more lua-esque.

In case it's better to just override `require`, how would yo approach returning preloaded modules?
Basically I don't know if luerl provides a mechanism to hook into require and if we can wrap it in this lib.

Thanks!

**Comment by Papipo** [comment]
Alternatively, creating a separate function (ie. `import`) that works like require but allows to return modules, that's also fine.

**Comment by Papipo** [comment]
This works, I can return anything from the imported script:

```elixir
defmodule ErluaTest do
  use ExUnit.Case
  doctest Erlua
  import Lua

  defmodule Custom do
    @imports %{
      "custom" => ~LUA"""
        return {
          greet = function(name)
            print("Printed hello, " .. name)
          end,
          template = function(name)
            return "Templating hello, " .. name
          end
      }
      """
    }

    def import([], [path], state) do
      {:ok, ret, state} = Luerl.New.do(state, @imports[path])
      {ret, state}
    end
  end

  test "Loading lua scripts dynamically" do
    entrypoint = ~LUA"""
      print("I am the entrypoint")
      local imported = import("custom")
      imported.greet("world")
      print("gonna print a rendered template: " .. imported.template("world"))
    """

    result =
      Lua.new()
      |> Lua.set!([:import], {Custom, :import, []})
      |> Lua.eval!(entrypoint)
  end
end
```

But as you can see I had to use Luerl `do`, you can't use `eval!` there.
Also, conceptually I don't know if the import function should reuse the state or use a blank slate. I like the latter to avoid unexpected side effects.

**Comment by davydog187** [comment]
Hello @Papipo, I'm still thinking about ways to do this. I'll get back to you

**Comment by iggi42** [comment]
I would be down to writing a PR for this, assuming I understand the issue correctly.

Just to be clear, you are looking for a function like `setup_require(lua_vm, [ api_modules ]) :: lua_vm`, where an `api_module` is an elixir module implementing the `Lua.Api` behaviour, correct?

```elixir
defmodule ExampleAPI do
  
  use Lua.API, scope: "example.api"

  deflua square(name), _state do
    square * square
   end
end
```

soo it can be used in lua
```lua
local example = require("example.api")
return example.square(2) -- return 4
```

**Comment by davydog187** [comment]
https://hexdocs.pm/lua/Lua.html#load_api/3 already exists for that purpose. I believe the ask is a bit different

**Comment by iggi42** [comment]
yeah,
you can already do that, there is no need to change this project for that, imo.

See [package.searchers](https://www.lua.org/manual/5.3/manual.html#pdf-package.searchers) in lua for how to.

**Comment by davydog187** [comment]
There's quite a bit of documentation I'd like to add on this topic. So ill close when that is done @Papipo
